sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsContainer.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.metrics.MetricUpdates.MetricUpdate;
+
+/**
+ * Holds all of the metrics produced for a single step and unit-of-commit.
+ *
+ * <p>A thread-local variable holds the {@link MetricsContainer} that should be used for any
+ * metric updates produced from the executing code by interactions with the user-facing metric
+ * interfaces (eg., {@link Counter}).
+ * *
+ * <p>For a given runner to support metrics it is currently necessary to do 3 things:
+ * <ol>
+ *   <li>Create a {@link MetricsContainer} for each scope that metrics will be reported at.
+ *   </li>
+ *   <li>Make sure to call {@link #setMetricsContainer}
+ *   </li>
+ *   <li>Use {@link #getCumulative()} to get {@link org.apache.beam.sdk.metrics.MetricUpdates}
+ *   representing all the metric changes and report/aggregate those appropriately.
+ *   </li>
+ *   <li>Optionally, use {@link #getDeltas()} to get only the changes to a metric since the last
+ *   time and report those periodically. After reporting deltas and before the next call to
+ *   {@link #getDeltas()} the runner should invoke {@link #commitDeltas} to indicate that those
+ *   deltas have been incorporated.
+ *   </li>
+ * </ol>
+ */
+@Experimental(Kind.METRICS)
+public class MetricsContainer {
+
+  private static final ThreadLocal<MetricsContainer> CONTAINER_FOR_THREAD =
+      new ThreadLocal<MetricsContainer>();
+  private final AtomicBoolean committedFirstDelta = new AtomicBoolean(false);
+
+  private final String stepName;
+
+  private MetricsMap<MetricName, CounterCell> counters =
+      new MetricsMap<MetricName, CounterCell>() {
+        @Override
+        protected CounterCell createInstance() {
+          return new CounterCell();
+        }
+      };
+
+  private MetricsMap<MetricName, DistributionCell> distributions =
+      new MetricsMap<MetricName, DistributionCell>() {
+        @Override
+        protected DistributionCell createInstance() {
+          return new DistributionCell();
+        }
+      };
+
+  /**
+   * Create a new {@link MetricsContainer} associated with the given {@code stepName}.
+   */
+  public MetricsContainer(String stepName) {
+    this.stepName = stepName;
+  }
+
+  /**
+   * Set the {@link MetricsContainer} for the current thread.
+   */
+  public static void setMetricsContainer(MetricsContainer container) {
+    CONTAINER_FOR_THREAD.set(container);
+  }
+
+  /**
+   * Clear the {@link MetricsContainer} for the current thread.
+   */
+  public static void unsetMetricsContainer() {
+    CONTAINER_FOR_THREAD.remove();
+  }
+
+  /**
+   * Return the {@link MetricsContainer} for the current thread.
+   */
+  public static MetricsContainer getCurrentContainer() {
+    MetricsContainer container = CONTAINER_FOR_THREAD.get();
+    if (container == null) {
+      throw new IllegalStateException("Must call setMetricsContainer before reporting metrics.");
```

Comment:
Metrics are not affinitized to a container at creation time. Whenever a metric is reported it is against the current container at that point in time. The error should be here since it indicates that the we're not in a properly configured execution thread.

Longer term, we'll need to provide ways of propagating the MetricsContainer for a thread to other threads that are executing that step, but that will need to go through some APIs to make sure it is safely handled (eg., all those threads need to complete before we can report the step successfully completed, etc.)